### PR TITLE
[language] Add new FixedPoint32 module to stdlib

### PR DESCRIPTION
**Pull request by bob-wilson**
This adds a fixed-point numeric type with 32 fractional bits. This can be used for things like interest rates, conversion rates, etc. Besides functions to create these fixed-point values, this provides basic truncating multiply and divide functions.

## Motivation

Add support for non-integer values in a fixed-point representation.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added some new tests for this module, including all the interesting corner cases that I could think of.

### CODE REVIEW

File: language/move-lang/stdlib/modules/fixedpoint32.move [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    public fun create_from_raw_value(value: u64): T {
+        T { value }
+    }
+
+    // Create a fixed-point value from a rational number specified by its
+    // numerator and denominator. This function is for convenience; it is also
+    // perfectly fine to create a fixed-point value by directly specifying the
+    // raw value.
+    public fun create_from_rational(numerator: u64, denominator: u64): T {
```
Comment (tzakian) [inline language/move-lang/stdlib/modules/fixedpoint32.move]: Do we want to check and error if the `denominator` is zero?

File: language/move-lang/tests/functional/fixedpoint32/test.move [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    let not_three = FixedPoint32::multiply_u64(9, copy f2); // 9 * 0.333...
+    // multiply_u64 does NOT round -- it truncates -- so values that
+    // are not perfectly representable in binary may be off by one.
+    0x0::Transaction::assert(not_three == 2, not_three);
+
+    // Try again with a fraction slightly larger than 1/3.
+    let f3 = FixedPoint32::create_from_raw_value(FixedPoint32::get_raw_value(copy f2) + 1);
+    let three = FixedPoint32::multiply_u64(9, copy f3);
+    0x0::Transaction::assert(three == 3, three);
+}
```
Comment (tzakian) [inline language/move-lang/tests/functional/fixedpoint32/test.move]: Not needed now since this isn't being run... but having div-by-zero tests to make sure they error the way they should would be good for both `divide_u64` and `create_from_rational`

File: language/move-lang/stdlib/modules/fixedpoint32.move [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	// Scale the numerator to have 64 fractional bits and the denominator
+	// to have 32 fractional bits, so that the quotient will have 32
+	// fractional bits.
+	let scaled_numerator = (numerator as u128) << 64;
+	let scaled_denominator = (denominator as u128) << 32;
+	// Divide. This cannot overflow.
+	let quotient = scaled_numerator / scaled_denominator;
+	T { value: (quotient as u64) }
```
Comment (tnowacki) [inline language/move-lang/stdlib/modules/fixedpoint32.move]: Does this cast fail if the number is too large?

File: language/move-lang/tests/functional/fixedpoint32/test.move (hunk 2 above)
Comment (tnowacki) [reply language/move-lang/tests/functional/fixedpoint32/test.move]: Isn't this being run? It's in the functional tests

Comment (tnowacki) [reply language/move-lang/tests/functional/fixedpoint32/test.move]: Probably wan't more edge-case tests like, u64 max used in several places

File: language/move-lang/stdlib/modules/fixedpoint32.move (hunk 3 above)
Comment (dariorussi) [reply language/move-lang/stdlib/modules/fixedpoint32.move]: it will in the runtime for what is worth

**Comment by dariorussi** [comment]
how are additions and subtractions performed with this model/implementation?
Also is this PR to have a conversation about the fixed point (which is great) or do we intend to commit this in stdlib?

File: language/move-lang/stdlib/modules/fixedpoint32.move (hunk 3 above)
Comment (bob-wilson) [reply language/move-lang/stdlib/modules/fixedpoint32.move]: I don't know what I was thinking when I wrote that comment -- probably that these were only 32 bit values. Of course this can overflow. The current behavior of integer casts is that they truncate, but @vgao1996 is planning to change that. In the meantime, I can add an explicit check here.

File: language/move-lang/stdlib/modules/fixedpoint32.move (hunk 1 above)
Comment (bob-wilson) [reply language/move-lang/stdlib/modules/fixedpoint32.move]: I could add a check and abort. Is that any better than just letting the VM report an ARITHMETIC_ERROR?

File: language/move-lang/tests/functional/fixedpoint32/test.move (hunk 2 above)
Comment (bob-wilson) [reply language/move-lang/tests/functional/fixedpoint32/test.move]: Yes, I need to add more tests. It didn't work at all when I tried it yesterday because the functional tests were using the IR version of the stdlib where the new module does not exist.

**Comment by bob-wilson** [comment]
> how are additions and subtractions performed with this model/implementation?

We could add functions for those if they are commonly used. In the meantime, you can just add and subtract the raw values. There isn't anything special needed there.

> Also is this PR to have a conversation about the fixed point (which is great) or do we intend to commit this in stdlib?

I'd like to commit it as soon as I can get the tests to work.

File: language/move-lang/stdlib/modules/fixedpoint32.move (hunk 1 above)
Comment (tzakian) [reply language/move-lang/stdlib/modules/fixedpoint32.move]: I think that we should have an error check at the Move level. It's personal taste, so feel free to ignore, but I always feel weary relying on a VM/execution to bubble up the right error.

**Comment by tzakian** [comment]
The switchover has landed! So you should be able to get these working once you get the cycles :)

**Comment by bob-wilson** [comment]
> The switchover has landed! So you should be able to get these working once you get the cycles :)

Yes, I got it working today and I'm in the process of writing more tests.

File: language/move-lang/stdlib/modules/fixedpoint32.move (hunk 1 above)
Comment (bob-wilson) [reply language/move-lang/stdlib/modules/fixedpoint32.move]: OK, personal taste: My taste is to eventually add a specification block with an aborts_if condition showing that the function is expected to abort when the denominator is zero.

**Comment by bob-wilson** [comment]
I've updated this with more tests and comments. I also added an explicit check for underflow in the create_from_rational function, since that seems like it could be a source of unexpected behavior that we can disallow. This is ready for review now.

**Comment by bob-wilson** [comment]
One more change: I switched to use Transaction::assert and to report an abort code of 16 for the various arithmetic errors here, after I discovered that the abort code of 10 is generally used for "insufficient balance" errors. After we change the VM to fail on overflow with integer type casts (and perhaps pending various discussions of coding style), I am planning to remove the explicit checks for overflows, which will leave only the one explicit Transaction::assert for underflow in create_from_rational.

**Comment by bob-wilson** [comment]
I updated this PR to take advantage of #2805 to report overflows in the type casts.

**Comment by vgao1996** [comment]
I know we probably don't have a use case for now, but do we want to offer something like `multiply_fixed32_fixed32` or `divide_fixed32_fixed32`?

**Comment by bob-wilson** [comment]
> I know we probably don't have a use case for now, but do we want to offer something like `multiply_fixed32_fixed32` or `divide_fixed32_fixed32`

I'd rather wait to see what is needed, but that's just my personal preference.
